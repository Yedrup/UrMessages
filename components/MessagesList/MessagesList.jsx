import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import SimpleBar from 'simplebar-react';
import MessageItem from '../Message/Message';

const SimpleBarStyled = styled(SimpleBar)`
  .simplebar-scrollbar:before {
    border-radius: 7px;
    left: 2px;
    right: 2px;
    opacity: 0;
    transition: opacity 0.2s linear;
    background: linear-gradient(${({ theme }) => theme.gradientToBottom});
  }
`;

const ListOfItemsStyled = styled.div`
  display: grid;
  grid-template-columns: 1fr;
  grid-auto-rows: auto;
  grid-gap: 40px;
`;

const MessagesList = ({ messages }) => {
  if (!messages?.length) return null;
  return (
    <SimpleBarStyled
      style={{ maxHeight: 900 }}
      autoHide={false}
      scrollbarMaxSize={650}
      scrollbarMinSize={425}
    >
      <ListOfItemsStyled data-testid="thread-message-list">
        {messages.map(message => {
          return <MessageItem message={message} key={message?.id} />;
        })}
      </ListOfItemsStyled>
    </SimpleBarStyled>
  );
};

export default MessagesList;

MessagesList.propTypes = {
  message: PropTypes.array,
};
