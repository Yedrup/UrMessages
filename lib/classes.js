import casual from 'casual-browserify';
import { v4 as uuidv4 } from 'uuid';
import formatDate from './utils/date';
// import faker from 'faker';

export class MessageClass {
  constructor({
    parentId = null,
    user = uuidv4(),
    isPublic = false,
    isThreadAllowed = true,
    users = null,
    title = casual.title,
    content = casual.words(),
  }) {
    const dateNow = Date.now();
    this.id = uuidv4();
    this.parentId = parentId;
    this.user = user;
    this.isPublic = isPublic;
    this.isThreadAllowed = isThreadAllowed;
    this.users = users;
    this.title = title;
    this.content = content;
    this.dataInMs = dateNow;
    this.date = formatDate(dateNow); // TODO:  remove it, dateInMs is enough and more flexible
  }
}
