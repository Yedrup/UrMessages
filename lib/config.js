export const API_ENDPOINT_DEV = `http://localhost:3000`;
