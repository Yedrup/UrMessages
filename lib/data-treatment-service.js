import {
  getAllMessagesAPI,
  getSingleMessageByIdAPI,
  getAllMessagesFromTypeAPI,
  postMessageAPI,
} from './api';

const getAllMessages = () => {
  const resp = getAllMessagesAPI();
  return resp;
};

const getAllMessagesFromType = ({ type }) => {
  const resp = getAllMessagesFromTypeAPI({ type });
  return resp;
};

const getSingleMessageById = ({ type, id }) => {
  const resp = getSingleMessageByIdAPI({ type, id });
  return resp;
};

const postMessage = ({ message }) => {
  const resp = postMessageAPI({ message });
  return resp;
};

export {
  getAllMessages,
  getSingleMessageById,
  getAllMessagesFromType,
  postMessage,
};
