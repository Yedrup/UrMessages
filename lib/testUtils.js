const fakePrivateMessage = {
  id: 'e0f17993-697b-4807-88b9-93e139396ddc',
  parentId: null,
  user: '1677f769-8fb5-41b1-aa42-9110379a1717',
  isPublic: false,
  isThreadAllowed: true,
  users: null,
  title: 'Molestias molestias',
  date: 'Sunday, August 2, 2020, 1:02:01 PM',
  content: 'reiciendis asperiores tempore enim earum velit consequatur',
};

const fakePublicMessage = {
  id: '5a707960-1c54-4bde-a5b9-c83a3318b24a',
  parentId: null,
  user: '91d979ae-2ea2-4c8b-932f-156299a2f8bb',
  isPublic: true,
  isThreadAllowed: true,
  users: null,
  title: 'Dolorem perferendis in',
  date: 'Sunday, August 2, 2020, 7:30:15 AM',
  content: 'vero architecto doloribus exercitationem aut veritatis eveniet',
};

const fakeMessageNoThreadAllowed = {
  id: '934badde862-ddb7-4b9d-83fe-555c8734be4e',
  user: '55ac9363-267a-40bd-93b0-23cdb7dcb390',
  parentId: 'cb15ac13-d1da-4e37-8045-35571fdc4e5d',
  isPublic: true,
  isThreadAllowed: false,
  title: 'Dolorem perferendis in',
  date: 'Sunday, August 2, 2020, 9:45:30 PM',
  content: 'vero architecto doloribus exercitationem aut veritatis eveniet',
};

const fakeMessages = [
  {
    id: '9f72b5e2-9cee-4003-8356-ad48745579be',
    parentId: null,
    user: 'e33c94de-8c4d-44d8-95bf-8d491a82846a',
    isPublic: true,
    isThreadAllowed: true,
    users: null,
    title: 'Architecto alias saepe',
    date: 'Sunday, August 2, 2020, 4:05:31 AM',
    content: 'ea aperiam quia nam commodi quam incidunt',
  },
  {
    id: '5a707960-1c54-4bde-a5b9-c83a3318b24a',
    parentId: null,
    user: '91d979ae-2ea2-4c8b-932f-156299a2f8bb',
    isPublic: true,
    isThreadAllowed: true,
    users: null,
    title: 'Dolorem perferendis in',
    date: 'Sunday, August 2, 2020, 7:30:15 AM',
    content: 'vero architecto doloribus exercitationem aut veritatis eveniet',
  },
];

const fakeThreadParent = {
  id: 'c4b1e9e2-0ca6-4a84-b6e5-538241f1b447',
  parentId: null,
  user: '1677f769-8fb5-41b1-aa42-9110379a1717',
  isPublic: false,
  isThreadAllowed: true,
  title: 'Lorem ipsum dolor sit amet consectetur',
  users: ['55ac9363-267a-40bd-93b0-23cdb7dcb390'],
  date: 'Sunday, August 2, 2020, 1:47:56 PM',
  content:
    'Doloremque tenetur dolore iure nisi quidem laudantium, hic earum necessitatibus',
  threadId: '29711c6d-fe5a-4908-8094-ba31e102b28d',
};

const fakeThread = {
  parentId: 'c4b1e9e2-0ca6-4a84-b6e5-538241f1b447',
  id: '29711c6d-fe5a-4908-8094-ba31e102b28d',
  isPublic: false,
  messages: [
    {
      id: '934bae862-bbb7-4b9d-83fe-555c8734be4e',
      parentId: '29711c6d-fe5a-4908-8094-ba31e102b28d',
      user: '1677f769-8fb5-41b1-aa42-9110379a1717',
      isPublic: false,
      isThreadAllowed: false,
      title: 'Delectus itaque dolores',
      date: 'Sunday, August 2, 2020, 9:35:52 PM',
      content: 'odio illo possimus veniam qui ratione nihil',
    },
  ],
};

const fakeThreadWithoutMessages = {
  parentId: 'c4b1e9e2-0ca6-4a84-b6e5-538241f1b447',
  id: '29711c6d-fe5a-4908-8094-ba31e102b28d',
  isPublic: false,
  messages: [],
};

export {
  fakeMessageNoThreadAllowed,
  fakePublicMessage,
  fakePrivateMessage,
  fakeMessages,
  fakeThread,
  fakeThreadParent,
  fakeThreadWithoutMessages,
};
