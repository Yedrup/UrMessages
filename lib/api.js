import fetch from 'isomorphic-fetch';
import { API_ENDPOINT_DEV } from './config';

const callAPI = async (url, opt = {}) => {
  // console.log('callAPI --- url', url);
  try {
    const response = await fetch(url, opt);
    if (response.ok) {
      const data = await response.json();
      // console.log('callAPI --- data', data);
      return data;
    }
    // create an error
    const error = new Error(response.statusText);
    error.customMessage = await response.text();
    throw error;
  } catch (error) {
    // console.log('callAPI --- error', error);
    throw error;
  }
};

const getAllMessagesAPI = () => {
  const resp = callAPI(`${API_ENDPOINT_DEV}/api/messages`);
  return resp;
};

const getAllMessagesFromTypeAPI = ({ type }) => {
  const resp = callAPI(`${API_ENDPOINT_DEV}/api/messages/${type}`);
  return resp;
};

const getSingleMessageByIdAPI = ({ type, id }) => {
  const resp = callAPI(`${API_ENDPOINT_DEV}/api/messages/${type}/${id}`, {});
  return resp;
};

const postMessageAPI = ({ message }) => {
  let messageType = message.isPublic ? 'public' : 'private';
  let resp;
  let isAThread = !!message.parentId;
  // remove the possibility to create a thread of a thread message so far
  if (isAThread) message.isThreadAllowed = false;
  const options = {
    method: 'POST',
    body: JSON.stringify(message),
    headers: {
      'Content-Type': 'application/json',
    },
  };

  if (!isAThread) {
    resp = callAPI(`${API_ENDPOINT_DEV}/api/messages/${messageType}`, options);
  } else {
    resp = callAPI(
      `${API_ENDPOINT_DEV}/api/messages/${messageType}/${message.id}`,
      options
    );
  }
  return resp;
};

export {
  getAllMessagesAPI,
  getSingleMessageByIdAPI,
  getAllMessagesFromTypeAPI,
  postMessageAPI,
};
