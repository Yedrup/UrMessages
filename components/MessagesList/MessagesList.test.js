import { render, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import MessagesList from './MessagesList';
import { fakeMessages } from '../../lib/testUtils';

afterEach(cleanup);
describe('<MessageList/>', () => {
  it('properly renders the component - snapshot', () => {
    const { asFragment } = render(<MessagesList messages={fakeMessages} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('It renders the number of given messages', () => {
    render(<MessagesList messages={fakeMessages} />);
    const messages = screen.getAllByTestId('message-item');
    expect(messages).toHaveLength(2);
  });

  it('Should not render the component returning null', () => {
    const { container } = render(<MessagesList messages={[]} />);
    expect(container).toBeEmptyDOMElement();
  });
});
