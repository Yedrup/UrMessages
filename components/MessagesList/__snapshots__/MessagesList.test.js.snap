// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList/> properly renders the component - snapshot 1`] = `
<DocumentFragment>
  <div
    class="MessagesList__SimpleBarStyled-epu0q5-0 bmTbbH"
    data-simplebar="init"
    style="max-height: 900px;"
  >
    <div
      class="simplebar-wrapper"
    >
      <div
        class="simplebar-height-auto-observer-wrapper"
      >
        <div
          class="simplebar-height-auto-observer"
        />
      </div>
      <div
        class="simplebar-mask"
      >
        <div
          class="simplebar-offset"
          style="right: 0px; bottom: 0px;"
        >
          <div
            class="simplebar-content-wrapper"
            style="height: auto; overflow-x: hidden; overflow-y: hidden;"
          >
            <div
              class="simplebar-content"
            >
              <div
                class="MessagesList__ListOfItemsStyled-epu0q5-1 fefGqq"
                data-testid="thread-message-list"
              >
                <article
                  class="Message__StyledMessageItem-uraoec-0 isHSvA"
                  data-testid="message-item"
                >
                  <header>
                    <h3>
                      <span>
                        Architecto alias saepe
                      </span>
                      <span>
                        ðŸ“£ 
                      </span>
                    </h3>
                    <div
                      class="user"
                    >
                      <div
                        class="user__image"
                      >
                        <div
                          class="Avatar__AvatarStyled-fet300-0 Avatar___StyledAvatarStyled-fet300-1 eecPIG c-avatar"
                        >
                          <img
                            alt="user undefined"
                            src="https://i.pravatar.cc/70px?img=44"
                          />
                        </div>
                      </div>
                      <span
                        class="user__name"
                      />
                    </div>
                  </header>
                  <section>
                    <p>
                      ea aperiam quia nam commodi quam incidunt
                    </p>
                  </section>
                  <footer>
                    <p>
                      Sunday, August 2, 2020, 4:05:31 AM
                    </p>
                    <a
                      class="c-button--text"
                      data-testid="thread-link"
                      href="/messages/public/9f72b5e2-9cee-4003-8356-ad48745579be"
                    >
                      SEE THREAD â†’
                    </a>
                  </footer>
                </article>
                <article
                  class="Message__StyledMessageItem-uraoec-0 isHSvA"
                  data-testid="message-item"
                >
                  <header>
                    <h3>
                      <span>
                        Dolorem perferendis in
                      </span>
                      <span>
                        ðŸ“£ 
                      </span>
                    </h3>
                    <div
                      class="user"
                    >
                      <div
                        class="user__image"
                      >
                        <div
                          class="Avatar__AvatarStyled-fet300-0 Avatar___StyledAvatarStyled-fet300-1 eecPIG c-avatar"
                        >
                          <img
                            alt="user undefined"
                            src="https://i.pravatar.cc/70px?img=8"
                          />
                        </div>
                      </div>
                      <span
                        class="user__name"
                      />
                    </div>
                  </header>
                  <section>
                    <p>
                      vero architecto doloribus exercitationem aut veritatis eveniet
                    </p>
                  </section>
                  <footer>
                    <p>
                      Sunday, August 2, 2020, 7:30:15 AM
                    </p>
                    <a
                      class="c-button--text"
                      data-testid="thread-link"
                      href="/messages/public/5a707960-1c54-4bde-a5b9-c83a3318b24a"
                    >
                      SEE THREAD â†’
                    </a>
                  </footer>
                </article>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="simplebar-placeholder"
        style="width: 0px; height: 0px;"
      />
    </div>
    <div
      class="simplebar-track simplebar-horizontal"
      style="visibility: hidden;"
    >
      <div
        class="simplebar-scrollbar simplebar-visible"
        style="width: 0px; display: none;"
      />
    </div>
    <div
      class="simplebar-track simplebar-vertical"
      style="visibility: hidden;"
    >
      <div
        class="simplebar-scrollbar simplebar-visible"
        style="height: 0px; display: none;"
      />
    </div>
  </div>
</DocumentFragment>
`;
